core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DecorrelateInnerUnnestWithGlobalAggregation.java

```diff
@@ -0,0 +1,365 @@
+import io.trino.sql.planner.plan.CorrelatedJoinNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.ProjectNode;
+import io.trino.sql.planner.plan.UnnestNode;
+import io.trino.sql.planner.plan.UnnestNode.Mapping;
+import io.trino.sql.tree.Expression;
+import io.trino.sql.tree.IsNotNullPredicate;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.trino.matching.Pattern.nonEmpty;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.sql.ExpressionUtils.and;
+import static io.trino.sql.planner.iterative.rule.AggregationDecorrelation.rewriteWithMasks;
+import static io.trino.sql.planner.iterative.rule.Util.restrictOutputs;
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isScalar;
+import static io.trino.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static io.trino.sql.planner.plan.AggregationNode.singleGroupingSet;
+import static io.trino.sql.planner.plan.JoinNode.Type.INNER;
+import static io.trino.sql.planner.plan.JoinNode.Type.LEFT;
+import static io.trino.sql.planner.plan.Patterns.CorrelatedJoin.correlation;
+import static io.trino.sql.planner.plan.Patterns.CorrelatedJoin.filter;
+import static io.trino.sql.planner.plan.Patterns.correlatedJoin;
+import static io.trino.sql.tree.BooleanLiteral.TRUE_LITERAL;
+
+/**
+ * This rule finds correlated UnnestNode in CorrelatedJoinNode's subquery and folds
+ * them into UnnestNode representing LEFT JOIN UNNEST.
+ * This rule transforms plans, where:
+ * - UnnestNode in subquery is based only on correlation symbols
+ * - UnnestNode in subquery is INNER without filter
+ * - subquery contains global aggregation over the result of unnest
+ * Additionally, other global aggregations, grouped aggregations and projections
+ * in subquery are supported.
+ * <p>
+ * Transforms:
+ * <pre>
+ * - CorrelatedJoin (LEFT or INNER) on true, correlation(c1, c2)
+ *      - Input (a, c1, c2)
+ *      - Aggregation
+ *           global grouping
+ *           agg <- agg1(x)
+ *           - Projection
+ *                x <- foo(y)
+ *                - Aggregation
+ *                     group by (g)
+ *                     y <- agg2(u)
+ *                     - Unnest INNER
+ *                          g <- unnest(c1)
+ *                          u <- unnest(c2)
+ *                          replicate: ()
+ * </pre>
+ * Into:
+ * <pre>
+ * - Projection (restrict outputs)
+ *      - Aggregation
+ *           group by (a, c1, c2, unique)
+ *           agg <- agg1(x) mask(mask_symbol)
+ *           - Projection
+ *                x <- foo(y)
+ *                - Aggregation
+ *                     group by (g, a, c1, c2, unique, mask_symbol)
```

Comment:
If we have unique assigned for `a, c1, c2` we do not have to put those in set of grouping columns. Just `unique` would be enough. Right? 
Not sure if this is any harm though?
Do you know if for execution-wise one is better than the other?

Reply:
Oh - never mind. This comment does not make sense. If they were not part of grouping set we would have to wrap them into `arbitrary` aggregation. Keep it.